Migrate to new timezone database

dickermoshe opened the pull request:
This PR introduces the following changes

1. The convert function, which is basically a way to get an epoch from a local datetime (jan 1st is new york is at what timestamp) has had it arguments moved to a year, month day... which are more in line how to  think about them and also remove the footgun of using a non-utc datetime when converting. 
2. Use a new timezone database provided by https://github.com/kshetline/tubular_time_tzdb which is more stable than it's dart counterpart.  This fixes a number a bugs and web compatibility. It also allows us to calculate the dst for times past the end date of the timezone database using dst rules.
3. Add an extensive Java test that rigorously check the results of timezone conversions.
4. A small refactor to make sugar support more timezone providers in the future.
5. `span` has been replaced by `offset`. This means users won't know the `abbreviation` of the current offset (`DST` or `EST`) or whether it is currently `isDst`. This is a change was made for the following reasons:
  a. Other timezone providers will have a hard time supporting this, there are very large disagreements what is considered `isDst` and in IMHO its best to avoid this issue completely.
  b. The span abbreviation would be really nice to have. We'll explore ways to bring it back. The issue is that DST rules don't have abbreviations associated with them.



---

- [x] Add java testing to CI

Pante commented on sugar/lib/src/time/zone/providers/universal/universal_provider.dart:
```diff
@@ -0,0 +1,707 @@
+///
+/// This provider uses a bundled timezone database to provide timezone
+/// information for all known timezones.
+class UniversalTimezoneProvider extends UnmodifiableMapBase<String, Timezone> {
```
Nitpick: I think `EmbeddedTimezoneProvider` better represents what this class does.

dickermoshe replied:
100%

Pante commented on sugar/lib/src/time/zone/providers/universal/universal_provider.g.dart:
```diff
@@ -0,0 +1,929 @@
+part of 'universal_provider.dart';
+
+/// The timezone database.
+const _tzdb = {
```
I think this generated file needs work. Ideally we should process/structure it even further, i.e. move the metadata to another map/structure, convert some of the raw data into something more directly usable at runtime.

dickermoshe replied:
I agree that we should move the metadata.

However, although the data there looks very complicated, it's actually just the raw data that comes out from the official IANA database with some numbers and durations are encoded with `base60` to save room. but otherwise, it's just an intelligently compressed tzdb.

If we want to make this more compresses we would be in effect doing the same thing twice.

1. Generating this JSON.
2. Parsing It.
3. Saving into a readable format
4. Parsing it again at runtime.

This would be quite foolish to write another format when a working one exists. I can work on the code to make it more understandable, but I disagree that because it's not _readable_ we should ruin a perfectly good format.

More lines of code is always worse than less lines of code. By using another encoding we would just be adding more complexity for little gain.

If you would like to read more about this format you can see here https://github.com/kshetline/tubular_time_tzdb#timezone-descriptions

dickermoshe replied:
We could do that, but it would make the file much bigger, on web this is a big concern.

We could create one provider for web which builds it at runtime, and another which creates it at compile time.

But let's not get ahead of ourselves, the parsing process for the entire database takes only 12ms.
(although after I refactor the code to make it easier to understand it may take a little longer)

dickermoshe replied:
Generating at compile time increase the bundle size by 6mb, which is quite large even for a mobile app.

I think lazy loading is a smarter way to go for all platforms. 
It takes less that 1ms to load a single time zone.

Seems like very worthy tradeoff

Pante commented on sugar/lib/src/time/zone/providers/universal/universal_provider.dart:
```diff
@@ -0,0 +1,707 @@
+// ignore_for_file: public_member_api_docs, sort_constructors_first
```
I think this file needs to be broken down into several smaller files, it's difficult to follow what's happening.

dickermoshe replied:
Sure

dickermoshe replied:
I'll refactor it to make it much easier to understand what going on.

Pante reviewed (changes requested):
I think generally we should try to process the raw data further at compile time. In addition, what is the tzupdate.jar used for?

dickermoshe commented:
I've added CI to test using Java.
I've also renamed the provider and refactored the db parsing from the actual timezone abject.

If you think this looks fine I'll add a generated Provider that loads everything at compile time

Pante commented on .github/workflows/sugar_unit_test.yaml:
```diff
@@ -34,12 +34,24 @@ jobs:
+        with:
+          distribution: "adopt"
+          java-version: "11.0.26+4"
+      - name: Set up Maven
+        uses: stCarolas/setup-maven@v5
```
I don't think we need a separate Maven action as set-up Java should already provide it.

dickermoshe replied:
It sadly does not.

Pante commented on .github/workflows/sugar_unit_test.yaml:
```diff
@@ -34,12 +34,24 @@ jobs:
+        uses: actions/setup-java@v4
+        with:
+          distribution: "adopt"
+          java-version: "11.0.26+4"
```
Is there a reason we need to build specifically against Java 11?

dickermoshe replied:
You can pin it against any version you want.
Java 11 is used on every version of android (except the latest one) so I chose to use that one.

Pante replied:
Fair enough

Pante reviewed (commented):
Generally LGTM! One quick question, but should we use microseconds or milliseconds as the most granular precision?

dickermoshe commented:
milliseconds is ok.

I did make one breaking change.
convert now takes a year,month day... looking object instead of a EpochMicroseconds.

dickermoshe commented:
Also, I will not be adding one that compiles the raw timezones because it bloats the binary massively for only a small improvement

Pante commented:
I'll add an automated action that formats pull requests later today/early tomorrow.

dickermoshe commented:
In this latest release, timezone providers never return a "Factory" database and return null instead.
However the `Timezone(...)` constructor will return the "Factory" if it can't find it

```dart
// Breaking
Timezone.timezoneProvider["Factory"] // returns null

// Non-Breaking
final tz = Timezone("Factory"); // No such timezone, return factory
final tz = Timezone("Pizza");  // No such timezone, return factory
```

I can make this non-breaking if you want

dickermoshe commented:
> I'll add an automated action that formats pull requests later today/early tomorrow.

There is tons of formatting that this repo need and doing so will turn this PR into a nightmare to review.
Make a independant formatting PR once this PR is merged

Pante commented:
Yeah that’s a good point, this repo was created long before we standardised formatting etc internally, I’ll probably do it after this PR.

I think we should return factory instead of null!

dickermoshe commented:
> I think we should return factory instead of null!

Sure

dickermoshe commented:
> I think we should return factory instead of null! 

Done.
I've also added a test that ensures that every timezone which worked in the previous version of sugar works in this new one.

dickermoshe commented:
CI should pass now. Ready for merging when (hopefully) passes

dickermoshe commented:
any update on this?

Pante commented:
@dickermoshe is the PR ready for review? Just noticed that it's not a draft anymore. If so, i'll start reviewing it later today.

dickermoshe commented:
Yup!

dickermoshe commented:
I've updating the description of the PR.
Please read and take note of the major changes

dickermoshe commented:
I'm closing this PR because it has ended up in a very confusing place.
Gonna make a new PR that is easier to review.

@Pante Please format the entire repo and I'll get started

Pante commented:
The codebase has been migrated.
